ribs/emitters/rankers.py

```diff
@@ -0,0 +1,368 @@
+    ribs.emitters.rankers.ObjectiveRanker
+    ribs.emitters.rankers.TwoStageObjectiveRanker
+    ribs.emitters.rankers.ImprovementRanker
+    ribs.emitters.rankers.TwoStageImprovementRanker
+    ribs.emitters.rankers.RankerBase
+"""
+
+from abc import ABC, abstractmethod
+import numpy as np
+
+from ribs._docstrings import DocstringComponents, _core_docs
+
+__all__ = [
+    "RandomDirectionRanker",
+    "TwoStageRandomDirectionRanker",
+    "ObjectiveRanker",
+    "TwoStageObjectiveRanker",
+    "ImprovementRanker",
+    "TwoStageImprovementRanker",
+    "RankerBase",
+]
+
+# Define common docstrings
+_args = _core_docs["args"]
+_returns = DocstringComponents(
+    dict(index_batch="""
+    A batch of indicies representing a ranking of the solutions"""))
+
+
+class RankerBase(ABC):
+    """Base class for rankers.
+
+    Every ranker has a :meth:`rank` method that returns a list of indices
+    that indicate how the solutions should be ranked and a :meth:`reset` method
+    that resets the internal state of the ranker
+    (e.g. in :class:`ribs.emitters.rankers._random_direction_ranker`).
+
+    Child classes are only required to override :meth:`rank`.
+    """
+
+    @abstractmethod
+    def rank(self, emitter, archive, solution_batch, objective_batch,
+             measure_batch, metadata, add_statuses, add_values):
+        # pylint: disable=missing-function-docstring
+        pass
+
+    # Generates the docstring for rank
+    rank.__doc__ = f"""
+Generates a batch of indicies that represents an ordering of ``solution_batch``.
+
+Args:
+{_args.emitter}
+{_args.archive}
+{_args.solution_batch}
+{_args.objective_batch}
+{_args.measure_batch}
+{_args.metadata}
+{_args.add_statuses}
+{_args.add_values}
+
+Returns:
+{_returns.indices}
+    """
+
+    def reset(self, emitter, archive):
+        # pylint: disable=missing-function-docstring
+        pass
+
+    # Generates the docstring for reset
+    reset.__doc__ = f"""
+Resets the internal state of the ranker.
+
+Args:
+{_args.emitter}
+{_args.archive}
+   """
+
+
+class ImprovementRanker(RankerBase):
+    # TODO implement
+    pass
+
+
+class TwoStageImprovementRanker(RankerBase):
+    """Ranks the solutions based on the improvement in the objective.
+
+    This ranker originates in `Fontaine 2020
+    <https://arxiv.org/abs/1912.02400>`_ in which it was referred to as
+    "Improvement Emitter".
+    This ranker ranks solutions that improves the archive more higher.
+    Moreover, new solutions are ranked above improved ones, which ranks above
+    ones that were not added to the archive.
+    """
+
+    def rank(self, emitter, archive, solution_batch, objective_batch,
+             measure_batch, metadata, add_statuses, add_values):
+        # New solutions sort ahead of improved ones, which sort ahead of ones
+        # that were not added. Note that lexsort sorts the values in ascending
+        # order, so we use numpy fancy indexing to reverse the sorted array.
```

Comment:
`::-1` is not fancy indexing. Also, it may be more readable to use `np.flip` https://numpy.org/doc/stable/reference/generated/numpy.flip.html